composer/cli/launcher.py

```diff
@@ -0,0 +1,141 @@
+                text=True,
+            )
+        processes.append(process)
+
+    return set(processes)
+
+
+def monitor_processes(processes: Set[subprocess.Popen]):
+    while len(processes) > 0:
+        for process in processes:
+            if process.poll() is None:
+                # the process is still running
+                continue
+            else:
+                # return code of 0 implies clean exit
+                # return code of -9 implies sigkill, presumably from cleanup_processes()
+                if process.returncode not in (0, -9):
+                    if process.stdout is None:
+                        output = ""
+                    else:
+                        output = process.stdout.read()
+
+                    if process.stderr is None:
+                        stderr = ""
+                    else:
+                        stderr = process.stderr.read()
+                    exc = subprocess.CalledProcessError(
+                        process.returncode,
+                        cmd=process.args,
+                        output=output,
+                        stderr=stderr,
+                    )
+                    error_msg = [
+                        "Error in subprocess",
+                        "----------Subprocess STDOUT----------",
+                        exc.output,
+                        "----------Subprocess STDERR----------",
+                        exc.stderr,
+                    ]
+                    logger.exception("\n".join(error_msg), exc_info=exc)
+                    sys.exit(process.returncode)
+                else:
+                    # exited cleanly
+                    processes.remove(process)
+                    break
+        time.sleep(1)
+
+
+def cleanup_processes(processes: Set[subprocess.Popen]):
+    for process in processes:
+        if process.returncode is None:
+            logger.info("Killing subprocess %s with SIGTERM", process.pid)
+            try:
+                os.killpg(process.pid, signal.SIGTERM)
+            except ProcessLookupError:
+                pass
+
+    current_time = datetime.datetime.now()
+    while datetime.datetime.now() - current_time < datetime.timedelta(seconds=5):
+        if all(process.returncode is not None for process in processes):
+            break
+        time.sleep(0.1)
+
+    for process in processes:
+        if process.returncode is None:
+            logger.error("Killing subprocess %s with SIGKILL", process.pid)
+            try:
+                os.killpg(process.pid, signal.SIGKILL)
+            except ProcessLookupError:
+                pass
+
+    if torch.distributed.is_initialized():
+        torch.distributed.destroy_process_group()
+
+
+def main():
+    args = parse_args()
+
+    processes = launch_processes(args.nproc, args.training_script, args.training_script_args)
+
+    try:
+        monitor_processes(processes)
+    finally:
+        cleanup_processes(processes)
```

Comment:
I suppose so - it's not strictly necessary because the entry point defined in `setup.py` knows about the `main` function, but I guess a user could try to do something like `python composer/cli/launcher.py ...`?

Reply:
I think because Nvidia examples allows you to execute it as a script it'd probably be good to allow that as well.  I'd also want for it to be a 
```
if __name__ == "__main__":
  exit(main())
```
where the exit code gets passed through the script if possible (sometimes its nice for other automations/tests)